blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+    for "mask", are forked from the "states" element of the of the
+    `states` of the previous layer.
+
+    The concatenated `states` of all layers is the `states` of the stack.
+    In order to avoid conflict, the names of the elements appearing in the
+    `states` of each layers are suffixed with a '_<n>' before combining
+    them. The '<n>' is the number of the layer starting from 0.
+    The same process is true for the `contexts` and `outputs` of the
+    stack.
+
+    Parameters
+    ----------
+    transitions : list
+        List of recurrent units to use in each layer.
+    prototype : :class:`.FeedForward`, optional
+        The transformation applied to each states of each layer  when it
+        is forked to the sequences of the next layer. By
+        default a :class:`.Linear` transformation with biases is used.
+
+    Notes
+    -----
+    See :class:`.BaseRecurrent` for more initialization parameters.
+
+    """
+    def __init__(self, transitions, prototype=None, **kwargs):
+        super(RecurrentStack, self).__init__(**kwargs)
+
+        for d, transition in enumerate(transitions):
+            transition.name += '_%d' % d
+        self.transitions = transitions
+
+        if prototype is None:
+            # By default use Linear (with bias) as a default fork.
+            # This overrides the bad default inside Fork which is without bias.
+            # Yes I know use_bias=True us the defaut, but it is a big deal so
+            # I write it down explicitly.
+            prototype = Linear(use_bias=True)
+        depth = len(transitions)
+        self.forks = [Fork(self.normal_inputs(d), prototype=prototype)
+                      for d in range(1, depth)]
+
+        self.children = self.transitions + self.forks
+
+        # Programmatically fix the parameters of the  recurrent decoration
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
+        for t in ["states", "contexts", "outputs"]:
+            v = ['%s_%d' % (s, d)
+                 for d, transition in enumerate(transitions)
+                 for s in getattr(transition.apply, t)]
+            setattr(self.apply, t, v)
+
+    def normal_inputs(self, d):
+        transition = self.transitions[d]
+        return [name for name in transition.apply.sequences
+                if 'mask' not in name]
+
+    def _push_allocation_config(self):
+        # Configure the forks that connect the "states" element in the `states`
+        # of one layer to the elements in the `sequences` of the next layer,
+        # excluding "mask".
+        # This involves `get_dim` requests
+        # to the transitions. To make sure that it answers
+        # correctly we should finish its configuration first.
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim("states")
```

Comment:
Using "states" is ugly but this specific name does not leave any chance for error. You cant same the same on states[0]. I can keep the "states" string as a class parameter that you can override. I can also generate a warning if "states" is not located in index 0 of self.transition.apply.states